### PR TITLE
Feature/two_qubit_irb

### PR DESCRIPTION
**Pull request by cdbf1**
Add two qubit functionality to the IRB routine.

### CODE REVIEW

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 1]
```diff
@@ -21,15 +21,248 @@
...
+
+    return _reduce_single_qubit_clifford_seq([random.choice(set_A), random.choice(set_B)])
+
+
+def random_two_qubit_clifford() -> cirq.CliffordGate:
+    """Choose a random two qubit clifford gate.
+
+    Returns:
+        The random clifford gate.
+    """
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:212]: Could be nice to link https://arxiv.org/abs/1402.4848 and/or add a few lines describing to save folks a trip to Cirq (although the link to Cirq's `qubit_characterizations.py` file above is also nice).

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 2]
```diff
@@ -99,71 +332,97 @@ class IRB(BenchmarkingExperiment[Union[IRBResults, RBResults]]):
...
     ) -> None:
         """Constructs an IRB experiment.
 
         Args:
-            interleaved_gate: The single qubit Clifford gate to measure the gate error of. If None
+            interleaved_gate: The Clifford gate to measure the gate error of. If None
                 then no interleaving is performed and instead vanilla Randomize benchmarking is
                 performed.
-            num_qubits: The number of qubits to experiment on
+            num_qubits: The number of qubits to experiment on. Must either be 1 or 2 but is ignored.
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:345]:
```suggestion
            num_qubits: The number of qubits to experiment on. Must either be 1 or 2 but is ignored
```

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 3]
```diff
@@ -99,71 +332,97 @@ class IRB(BenchmarkingExperiment[Union[IRBResults, RBResults]]):
...
+        interleaved_gate: cirq.Gate | None = cirq.Z,
+        num_qubits: int | None = 1,
+        clifford_op_gateset: cirq.CompilationTargetGateset = cirq.CZTargetGateset(),
     ) -> None:
         """Constructs an IRB experiment.
 
         Args:
-            interleaved_gate: The single qubit Clifford gate to measure the gate error of. If None
+            interleaved_gate: The Clifford gate to measure the gate error of. If None
                 then no interleaving is performed and instead vanilla Randomize benchmarking is
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:343]:
```suggestion
                then no interleaving is performed and instead vanilla randomized benchmarking is
```

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 4]
```diff
@@ -99,71 +332,97 @@ class IRB(BenchmarkingExperiment[Union[IRBResults, RBResults]]):
...
+            clifford_op_gateset: The gateset to use when implementing the clifford operations.
+                Defaults to the CZ/GR set.
         """
-        if num_qubits != 1:
+        if interleaved_gate is not None:
+            num_qubits = interleaved_gate.num_qubits()
+        if num_qubits not in [1, 2]:
             raise NotImplementedError(
-                "IRB experiment is currently only implemented for single qubit use"
+                "IRB experiment is currently only implemented for single or two qubit qubit use."
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:354]:
```suggestion
                "IRB experiment is currently only implemented for single or two qubit use."
```

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 5]
```diff
@@ -99,71 +332,97 @@ class IRB(BenchmarkingExperiment[Union[IRBResults, RBResults]]):
...
+        super().__init__(num_qubits=num_qubits)
 
-        self.interleaved_gate = interleaved_gate
-        """The gate being interleaved"""
+        if interleaved_gate is not None:
+            self.interleaved_gate: cirq.CliffordGate | None = cirq.CliffordGate.from_op_list(
+                [interleaved_gate(*self.qubits)], self.qubits
+            )
+            self._interleaved_op = interleaved_gate(*self.qubits)
+            """The gate being interleaved"""
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:363]:
```suggestion
            """The operation being interleaved"""
```

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 6]
```diff
@@ -99,71 +332,97 @@ class IRB(BenchmarkingExperiment[Union[IRBResults, RBResults]]):
...
-        return cur
+        if self.num_qubits == 1:
+            return random_single_qubit_clifford()
+        return random_two_qubit_clifford()
 
-    @classmethod
-    def _random_single_qubit_clifford(cls) -> cirq.ops.SingleQubitCliffordGate:
-        """Choose a random single qubit clifford gate.
+    def gates_per_clifford(self, samples: int = 500) -> dict[str, float]:
+        """Samples a number of random Clifford operation and calculates the average number of
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:398]:
```suggestion
        """Samples a number of random Clifford operations and calculates the average number of
```

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 7]
```diff
@@ -99,71 +332,97 @@ class IRB(BenchmarkingExperiment[Union[IRBResults, RBResults]]):
...
+            ).item(),
+            "two_qubit_gates": np.mean(
+                [
+                    sum(1 for op in circuit.all_operations() if len(op.qubits) == 2)
+                    for circuit in sample
+                ]
+            ).item(),
+        }
 
     def _invert_clifford_circuit(self, circuit: cirq.Circuit) -> cirq.Circuit:
```
Comment (dowusu-antwi) [inline supermarq-benchmarks/supermarq/qcvv/irb.py:427]: Perhaps rename to something like:
```suggestion
    def _append_inverse_clifford_circuit(self, circuit: cirq.Circuit) -> cirq.Circuit:
```

**Review by dowusu-antwi — COMMENTED** [review]
More nits, mostly:

File: docs/source/apps/supermarq/qcvv/qcvv_irb_css.ipynb [hunk 8]
```diff
@@ -4,25 +4,27 @@
    "cell_type": "markdown",
```
Comment (dowusu-antwi) [inline docs/source/apps/supermarq/qcvv/qcvv_irb_css.ipynb:4]: <p>trues -&gt; true: "We see that the measured two qubit gate error is in good agreement with the true value used by &nbsp;the simulator."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Infleqtion/client-superstaq/pull/1068/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

File: supermarq-benchmarks/supermarq/qcvv/irb.py [hunk 9]
```diff
@@ -21,15 +21,248 @@
...
+    # X
+    cirq.CliffordGate.from_clifford_tableau(
+        cirq.CliffordTableau(
+            1,
+            rs=np.array([False, False], dtype=np.dtype("bool")),
+            xs=np.array([[True], [False]], dtype=np.dtype("bool")),
+            zs=np.array([[False], [True]], dtype=np.dtype("bool")),
+            initial_state=0,
+        )
+    ),
```
[the remark this answers is not in the document]
Comment (cdbf1) [reply supermarq-benchmarks/supermarq/qcvv/irb.py:86]: Good catch, I miss-copied the table. Thanks!

File: supermarq-benchmarks/supermarq/qcvv/irb.py (hunk 7 above)
Comment (cdbf1) [reply supermarq-benchmarks/supermarq/qcvv/irb.py:425]: This function isn't actually used any more so I've removed it.

**Review by dowusu-antwi — APPROVED** [review]
lgtm! (after branch update)

**Pull request merged**